src/index.ts

```diff
@@ -7,8 +7,34 @@ export type Unionized<Record, TaggedRecord> = {
   match: Match<Record, TaggedRecord>
 } & Creators<Record, TaggedRecord>
 
+/*
+ * Helper types for differentiating empty parameter types ({} and null)
+ */
+export type Bool = 'T' | 'F'
+
+export type If<B extends Bool, Then, Else> = {
+  T: Then
+  F: Else
+}[B]
+
+// this is a necessary intermediate type, can't be merged into ObjectHasKeys<O>
+export type StringContains<S extends string, L extends string> = (
+  { [K in S]: { _: 'T' } } &
+  { [key: string]: { _: 'F' } }
+)[L]['_']
+
+export type ObjectHasKeys<O> = StringContains<keyof O, keyof O>
```

Comment:
How about just using `typelevel-ts` for this machinery? There's really no downside since it's purely a static dependency.

Reply:
This is actually different from `typelevel-ts`'s implementation of `StringContains`, which fails when applied to the keys of a primitive like `number` or `boolean`.

I can look into trying to work around that failure, or perhaps getting a PR into `typelevel-ts`, since it would definitely be preferable to depend on an external library for all of this.

Reply:
I see. Well, no need to go into contortions for it.